src/AzureAppConfigurationImpl.ts

```diff
@@ -299,9 +377,25 @@ export class AzureAppConfigurationImpl implements AzureAppConfiguration {
         }
     }
 
+    async #refreshFeatureFlags(): Promise<void> {
+        // if still within refresh interval/backoff, return
+        if (!this.#featureFlagRefreshTimer.canRefresh()) {
+            return Promise.resolve();
+        }
+
+        try {
+            await this.#loadFeatureFlags();
+            this.#featureFlagRefreshTimer.reset();
+        } catch (error) {
+            // if refresh failed, backoff
+            this.#featureFlagRefreshTimer.backoff();
+            throw error;
+        }
+    }

```

Comment:
Do we need to run callbacks after successful refresh of feature flags

Reply:
no we don't, as we don't have an similar `onRefresh()` api compared with configuration setting refresh.

Reply:
Should we add that API? How can we enable users to run some custom code when feature flag refresh succeeds?

Reply:
> Should we add that API? 

No idea, depends on whether there is valid custom requests.
Since `refresh()` returns a promise, one can always use `refresh().then(successCallback).catch(errorCallback)` to run custom code.
Currently public APIs `refresh()` is shared between ff & kv, and `onRefresh()` is dedicated to kv (based on current descrption). A little bit inconsistent as you mentioned. 

 So two options in my mind, I prefer a) for simplicity and not over-optimizing it.
- a) `refresh()` and `onRefresh()` both shared between ff&kv. I.e. we run callbacks after successfully refresh **BOTH** kv and ff.
- b) add a new API like `onFeatureFlagRefresh()`, (and a new `refreshFeatureFlags()` which is consistent but less easy to use) if we want to focus on ff refresh, and run callbacks even if kv refresh fails. It sounds too higher granularity, and I don't know if that's necessary.



For reference, just curious about what's the case in dotnet provider.

Reply:
We dont provide any special way to register config change callbacks, but .NET has a way for users to implement [options pattern](https://learn.microsoft.com/en-us/dotnet/core/extensions/options#ioptionsmonitor) to get notified of configuration changes. This is triggered every time any config value changes (regardless of kv of ff). So I like your first option too. BTW, is there a way for users to tell what was refreshed - kv or ff?

Reply:
> , is there a way for users to tell what was refreshed - kv or ff?

So far, for option a), no. The experience would be: you provide custom callbacks either via `refresh().then()` or `onRefresh()`, and after `refresh()` is successful (meaning both ff and kv are refreshed), the callback is executed, without knowing which kv of ff is actually updated. 
The design was because you can always get the up-to-date kv/ff by your own in the callback, and we don't compare the new/old values for you. If you are interested in the changed ones, you can always keep a cache and compare by your own.

Reply:
That's reasonable. We should also plan to add logging statements to provide details of refresh events in future. For this PR, option A sounds good.

Reply:
> add logging statements to provide details of refresh events in future

tracked in #75

Reply:
99f2c29

`refresh()` and `onRefresh()` now apply for both key-values and feature flags. `refresh()` will try refreshing both kv and ff. If either kv of ff is successfully refreshed, the callbacks registered via `onRefresh()` will be called.

Pending task is to refresh feature flags using page etag.

Reply:
4ec140a : refresh feature flags with page etag check first.. 

Providing page etags on listing configs, if all pages from all selectors return 304, then there is no need to refresh. 
If any page returns 200, we short-circuit the loop and load all feature flags from scratch, do the composition, and update pageEtags.